### PR TITLE
gh-80793: Edit unittest.rst on SetUpClass calls

### PR DESCRIPTION
**Pull request by vrpolakatcisco**
Signed-off-by: Vratko Polak <vrpolak@cisco.com>

* Issue: gh-80793

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by vrpolakatcisco** [comment]
> Our records indicate we have not received your CLA.

Signed, but waiting for the records to get updated.

**Comment by vrpolakatcisco** [comment]
> bedevere/news — News entry found in Misc/NEWS.d 

I fail at blurb. How can I fix?

**Comment by vrpolakatcisco** [comment]
> waiting for the records to get updated

https://check-python-cla.herokuapp.com says it is updated now.

File: Misc/NEWS.d/next/Documentation/2019-04-12-12-30-08.bpo-36612.8AGXVv.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+skip news
```
Comment (MakDon) [inline Misc/NEWS.d/next/Documentation/2019-04-12-12-30-08.bpo-36612.8AGXVv.rst]:
```suggestion
```

**Review by MakDon — CHANGES_REQUESTED** [review]

**Comment by MakDon** [comment]
Hi, @vrpolakatcisco ,it is no need to add a 'skip news' in  Misc/NEWS.d. A core dev would review your commit and add a 'skip news' tag on the PR.
I haven't checked the document yet. I will check it later.

**Comment by MakDon** [comment]
Sorry, it seems that the workflow has change because i found the 'skip news' is commit by a bot.

File: Doc/library/unittest.rst [hunk 2]
```diff
@@ -2274,9 +2274,12 @@ These must be implemented as class methods::
         def tearDownClass(cls):
             cls._connection.destroy()
 
-If you want the ``setUpClass`` and ``tearDownClass`` on base classes called
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are empty.
+
+If you want the ``setUpClass`` and ``tearDownClass`` on a superclass
```
Comment (MakDon) [inline Doc/library/unittest.rst]:
```suggestion
If you want the ``setUpClass`` and ``tearDownClass`` on a base class
```
Is it a typo? Because your overriden method is the method on a superclass,  so maybe it is something wrong that 'if you want the method on a superclass called from your superclass method'

Comment (vrpolakatcisco) [reply Doc/library/unittest.rst]: I found a simpler way to make it clear normal inheritance rules apply, without referring to my original use case (chain of three classes).

**Comment by vrpolakatcisco** [comment]
> vrpolakatcisco wants to merge 2 commits into python:master from vrpolakatcisco:fixtures_doc

Well, my plan was to do a clean rebase. Not sure what to do with the "sync" merge commit.

File: Doc/library/unittest.rst [hunk 3]
```diff
@@ -2234,8 +2234,8 @@ Class and Module Fixtures
 
 Class and module level fixtures are implemented in :class:`TestSuite`. When
 the test suite encounters a test from a new class then :meth:`tearDownClass`
-from the previous class (if there is one) is called, followed by
-:meth:`setUpClass` from the new class.
+from the previous class (if it defines or inherits one) is called,
```
Comment (JulienPalard) [inline Doc/library/unittest.rst]: I don't think "defines or inheites" is clearer. What was your intent?

File: Doc/library/unittest.rst [hunk 4]
```diff
@@ -2269,7 +2269,7 @@ are a framework author it may be relevant.
 setUpClass and tearDownClass
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-These must be implemented as class methods::
+These must be implemented (or inherited) as class methods::
```
Comment (JulienPalard) [inline Doc/library/unittest.rst]: Same, what are you trying to clarify? As said in b.p.o normal inheritance rules apply. I think we should be explicit about inheritance only when "anormal rule apply", not every time "normal" rules apply.

File: Doc/library/unittest.rst [hunk 5]
```diff
@@ -2282,9 +2282,8 @@ These must be implemented as class methods::
         def tearDownClass(cls):
             cls._connection.destroy()
 
-If you want the ``setUpClass`` and ``tearDownClass`` on base classes called
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The usual inheritance rules apply.
```
Comment (JulienPalard) [inline Doc/library/unittest.rst]: Same here, let's not be explicit about this being normal (if we're explicit here about the inheritance working normally, why not adding it everywhere else on the doc? No.)

**Review by JulienPalard — CHANGES_REQUESTED** [review]
Thanks for trying to enhance the documentation! I propose to simplify further, as it's only normal rules.

File: Doc/library/unittest.rst (hunk 3 above)
Comment (vrpolakatcisco) [reply Doc/library/unittest.rst]: > What was your intent?

I am just trying to win an argument by updating opponent's quote.
I spotted an edit [0] that I considered to me unnecessary, then started a discussion [1] on a similar edit, and my attempt to undo [2] such edits has been flatly refused.

[0] https://gerrit.fd.io/r/c/vpp/+/17092/3/test/test_bfd.py
[1] https://gerrit.fd.io/r/c/vpp/+/18579/1/test/test_sparse_vec.py#14
[2] https://gerrit.fd.io/r/c/vpp/+/18820

File: Doc/library/unittest.rst (hunk 4 above)
Comment (vrpolakatcisco) [reply Doc/library/unittest.rst]: The original formulation
  "These must be implemented as class methods"
sounds like normal rules do not apply for some reason, and each class needs to have an implementation (even if it only calls ancestor implementation).

File: Doc/library/unittest.rst (hunk 5 above)
Comment (vrpolakatcisco) [reply Doc/library/unittest.rst]: The original formulation
  "you must call up to them yourself"
once again sounds like normal rules do not apply for some reason, and you have to make the call explicit, instead of relying on inheritance.

> if we're explicit here about the inheritance working normally, why not adding it everywhere else on the doc?

I agree that confirming something is normal creates doubts on other (not explicitly confirmed things).

I am going to propose different edits, focusing on just removing the "wrong sounding" formulations.

File: Doc/library/unittest.rst [hunk 6]
```diff
@@ -2309,9 +2309,9 @@ Class and Module Fixtures
 -------------------------
 
 Class and module level fixtures are implemented in :class:`TestSuite`. When
-the test suite encounters a test from a new class then :meth:`tearDownClass`
-from the previous class (if there is one) is called, followed by
```
Comment (JulienPalard) [inline Doc/library/unittest.rst:2365]: Why removing the "if there is one"? It was right:
```
tearDownClass = getattr(previousClass, 'tearDownClass', None)
if tearDownClass is not None:
    ...
```

I don't see any problem with this paragraph, and it looks not correlated with bpo-36612.

File: Doc/library/unittest.rst [hunk 7]
```diff
@@ -2345,7 +2345,7 @@ are a framework author it may be relevant.
 setUpClass and tearDownClass
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-These must be implemented as class methods::
```
Comment (JulienPalard) [inline Doc/library/unittest.rst]: Looks unrelated to the issue, and I don't see how the new wording helps.

File: Doc/library/unittest.rst [hunk 8]
```diff
@@ -2372,7 +2370,7 @@ instead of as an error.
 setUpModule and tearDownModule
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-These should be implemented as functions::
```
Comment (JulienPalard) [inline Doc/library/unittest.rst]: I prefer the original wording.

File: Doc/library/unittest.rst [hunk 9]
```diff
@@ -2358,9 +2358,7 @@ These must be implemented as class methods::
         def tearDownClass(cls):
             cls._connection.destroy()
 
-If you want the ``setUpClass`` and ``tearDownClass`` on base classes called
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are present, but empty.
```
Comment (JulienPalard) [inline Doc/library/unittest.rst:2416]: This sentence is right, the implementations are present but empty. But now without context I don't see the point of mentioning it, why not removing it all?

**Review by JulienPalard — CHANGES_REQUESTED** [review]

File: Doc/library/unittest.rst (hunk 3 above)
Comment (JulienPalard) [reply Doc/library/unittest.rst]: I see you're trying to win an argument, but I stil not understand how modifing this specific sentence would have helped.

File: Doc/library/unittest.rst (hunk 4 above)
Comment (JulienPalard) [reply Doc/library/unittest.rst]: > sounds like normal rules do not apply for some reason

Yes, it tells it's not a method but a class method, it reminds you not to forgot to add the `@classmethod`, the original wording is OK.

> and each class needs to have an implementation (even if it only calls ancestor implementation)

Can you proove it with an example?

File: Doc/library/unittest.rst (hunk 5 above)
Comment (JulienPalard) [reply Doc/library/unittest.rst]: > once again sounds like normal rules do not apply for some reason

I agree.

> and you have to make the call explicit, instead of relying on inheritance

Inheritance never "call implicitly", as long as you re-implement the method, you *have* to call the parent implementation manually. But if you don't implement a method, the parent one is just inherited and no manual actions are needed. Only normal rules here.

File: Doc/library/unittest.rst (hunk 8 above)
Comment (vrpolakatcisco) [reply Doc/library/unittest.rst]: Not sure why module functions have "should" and class methods have "must", but I think I can make do edits also without affecting these two lines.

File: Doc/library/unittest.rst (hunk 9 above)
Comment (vrpolakatcisco) [reply Doc/library/unittest.rst:2416]: I will expand in next version.

File: Doc/library/unittest.rst [hunk 10]
```diff
@@ -2309,17 +2309,20 @@ Class and Module Fixtures
...
-``tearDownModule`` from the previous module is run, followed by
+Similarly if a test is from a different module from the previous test, then
+``tearDownModule`` from the previous module is called, followed by
 ``setUpModule`` from the new module.
 
 After all the tests have run the final ``tearDownClass`` and
 ``tearDownModule`` are run.
 
+The abovementioned calls are only executed,
+if the corresponding method or function is found.
```
Comment (vrpolakatcisco) [inline Doc/library/unittest.rst:2376]: Choosing "method is found", instead of "method exists" or "class has a method", in order to hint method resolution order is in effect.

File: Doc/library/unittest.rst [hunk 11]
```diff
@@ -2358,9 +2361,10 @@ These must be implemented as class methods::
         def tearDownClass(cls):
             cls._connection.destroy()
 
-If you want the ``setUpClass`` and ``tearDownClass`` on base classes called
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are present, but empty.
+That means the implemenation is executed for classes descending from
```
Comment (vrpolakatcisco) [inline Doc/library/unittest.rst:2417]: Choosing "descending" instead of "derived" or "child", to suggest the inheritance graphs can be complicated.

File: Doc/library/unittest.rst [hunk 12]
```diff
@@ -2358,9 +2361,10 @@ These must be implemented as class methods::
...
+:class:`TestCase`, but the implementation does not do anything,
```
Comment (vrpolakatcisco) [inline Doc/library/unittest.rst:2418]: Empty implementation executed means no special handling is there to interfere with normal inheritance rules.

File: Doc/library/unittest.rst (hunk 9 above)
Comment (JelleZijlstra) [reply Doc/library/unittest.rst:2416]: Honestly this still doesn't sound very useful to me. The new text doesn't really answer the question in the original bug report.

**Comment by willingc** [comment]
This PR has been languishing. I'm going to go ahead and close the PR and associated issue.

Thanks @vrpolakatcisco for the report and the PR.
